### PR TITLE
fossid-webapp: Display the version of FossID server

### PR DESCRIPTION
**Pull request by nnobelis**
Signed-off-by: Nicolas Nobelis <nicolas.nobelis@bosch.io>

### CODE REVIEW

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 1]
```diff
@@ -145,13 +145,47 @@ class FossId(
...
+
+        val response = service.getLoginPage()
+
+        response.use {
+            it.charStream().buffered().apply {
+                this.useLines { lines ->
+                    lines.forEach { line ->
+                        val matcher = regex.matchEntire(line)
+                        if (matcher != null) {
+                            if (matcher.groupValues.size == 2) {
```
Comment (mnonnenmacher) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:175]: Use `matcher != null && matcher.groupValues.size == 2` to reduce nesting.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 2]
```diff
@@ -145,13 +145,47 @@ class FossId(
...
+                                val version = matcher.groupValues[1]
+                                FossId.log.info { "Version from FossId Server is $version." }
+                                return version
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        log.warn { "Version from FossId Server cannot be found !" }
```
Comment (mnonnenmacher) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:185]: Remove whitespace before exclamation mark.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 3]
```diff
@@ -145,13 +145,47 @@ class FossId(
...
         options.mapValues { (k, v) ->
             v.takeUnless { k in secretKeys }.orEmpty()
         }
 
+    /**
+     * Extract the version version from the login page.
+     * Example: &nbsp;&nbsp;&nbsp;cli.  3.1.16 (build 5634934d, RELEASE)
+     */
+    private suspend fun parseVersion(): String? {
+        // TODO: replace with an API call when FossID provides a function (starting at version 21.2).
```
Comment (mnonnenmacher) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:164]: Is there a public ticket for this or release notes which we could reference here?

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt (hunk 3 above)
Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:164]: No, no publicly available information.

Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:164]: Should I remove the upcoming version estimate for the feature ?

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt (hunk 1 above)
Comment (sschuberth) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt]: Or even shorter: `matcher?.groupValues.size == 2`

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt (hunk 3 above)
Comment (sschuberth) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:164]: IMO no, it's still valuable to know, even if without reference.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 4]
```diff
@@ -145,13 +145,45 @@ class FossId(
...
+        val regex = Regex("^.*&nbsp;(cli. *[0-9\\. ]+\\(build[\\w, ]+\\)).*$")
+
+        val response = service.getLoginPage()
+
+        response.use {
+            it.charStream().buffered().apply {
+                this.useLines { lines ->
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:171]: Why the `apply` / `this` and not just `it.charStream().buffered().useLines {`?

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 5]
```diff
@@ -145,13 +145,45 @@ class FossId(
...
             v.takeUnless { k in secretKeys }.orEmpty()
         }
 
+    /**
+     * Extract the version version from the login page.
+     * Example: &nbsp;&nbsp;&nbsp;cli.  3.1.16 (build 5634934d, RELEASE)
+     */
+    private suspend fun parseVersion(): String? {
+        // TODO: replace with an API call when FossID provides a function (starting at version 21.2).
+        val regex = Regex("^.*&nbsp;(cli. *[0-9\\. ]+\\(build[\\w, ]+\\)).*$")
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:165]: See the IDE hint: No escaping required in `0-9\\.`.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 6]
```diff
@@ -145,13 +145,45 @@ class FossId(
...
+
+        val response = service.getLoginPage()
+
+        response.use {
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:169]: I'm wondering whether this outer `use` on the response body is even needed, as all that closing the response does is to close the body, which probably is what closing the buffered stream below already does. Maybe you can verify by debugging?

Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:169]: I was just wondering the same :)
Yes `useLines` closes the `BufferedReader` even if early returned.

Comment (sschuberth) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt]: And does the `BufferedReader` also close the underlying response body?

**Review by sschuberth — APPROVED** [review]

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt (hunk 6 above)
Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt]: Yes, I tested it. The `Source` deep within the `ResponseBody` is also closed.

**Comment by sschuberth** [comment]
Merging despite the unrelated `PipFunTest` failure.

**Pull request merged**